File: ios/Sources/RisemaxiGraft/Classes/ChannelRelease.swift
import Foundation

/// One entry of a channel document. Every value here is an unverified hint used to decide which
/// manifest to fetch; the manifest itself carries the signed copies that are enforced.
public struct ChannelRelease: Decodable {
    let id: String
    let counter: Int
    let rollout: Int
    let minNativeBuild: Int
    let manifest: String
    let sig: String

    private enum CodingKeys: String, CodingKey {
        case id, counter, rollout, minNativeBuild, manifest, sig
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decode(String.self, forKey: .id)
        counter = try container.decode(Int.self, forKey: .counter)
        rollout = try container.decode(Int.self, forKey: .rollout)
        guard rollout >= 0 && rollout <= 100 else {
            throw DecodingError.dataCorruptedError(forKey: .rollout, in: container, debugDescription: "rollout must be between 0 and 100: \(rollout)")
        }
        minNativeBuild = try container.decode(Int.self, forKey: .minNativeBuild)
        manifest = try container.decode(String.self, forKey: .manifest)
        sig = try container.decode(String.self, forKey: .sig)
    }
}

File: android/src/main/java/com/risemaxi/graft/classes/ChannelRelease.java
package com.risemaxi.graft.classes;

import androidx.annotation.NonNull;
import org.json.JSONException;
import org.json.JSONObject;

/**
 * One entry of a channel document. Every value here is an unverified hint used to decide which
 * manifest to fetch; the manifest itself carries the signed copies that are enforced.
 */
public class ChannelRelease {

    @NonNull
    private final String id;

    private final long counter;
    private final int rollout;
    @NonNull
    private final String nativeFingerprint;

    @NonNull
    private final String manifest;

    @NonNull
    private final String signature;

    public ChannelRelease(@NonNull JSONObject json) throws JSONException {
        this.id = json.getString("id");
        this.counter = json.getLong("counter");
        this.rollout = json.getInt("rollout");
        this.nativeFingerprint = json.getString("nativeFingerprint");
        this.manifest = json.getString("manifest");
        this.signature = json.getString("sig");
        if (rollout < 0 || rollout > 100) {
            throw new JSONException("rollout must be between 0 and 100: " + rollout);
        }
    }

    @NonNull
    public String getId() {
        return id;
    }

    public long getCounter() {
        return counter;
    }

    public int getRollout() {
        return rollout;
    }

    @NonNull
    public String getNativeFingerprint() {
        return nativeFingerprint;
    }

    @NonNull
    public String getManifest() {
        return manifest;
    }

    @NonNull
    public String getSignature() {
        return signature;
    }
}
